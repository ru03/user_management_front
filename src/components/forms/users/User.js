import React, { useEffect } from 'react';
import { Box, Button, Checkbox, FormControlLabel, Grid, TextField } from '@material-ui/core';

const User = ({ errors, handleChange, handleSubmit, isSubmitting, isResetForm, isUpdate, resetForm, touched, values }) => {
  useEffect(() => {
    if (isResetForm) resetForm();
  }, [isResetForm, resetForm]);
  return (
    <form onSubmit={handleSubmit}>
      <Box mt='20px'>
        <Grid container alignContent="center">
          <Grid xs={12} item>
            <TextField
              id="name"
              label="Name"
              fullWidth
              onChange={handleChange}
              value={values.name}
              error={touched.name && errors.name ? true : false}
              helperText={touched.name && errors.name}
            />
          </Grid>
          <Grid xs={12} item>
            <TextField
              id="lastName"
              label="Last Name"
              fullWidth
              onChange={handleChange}
              value={values.lastName}
              error={touched.lastName && errors.lastName ? true : false}
              helperText={touched.lastName && errors.lastName}
            />
          </Grid>
          <Grid xs={12} item>
            <TextField
              id="email"
              label="Email"
              fullWidth
              onChange={handleChange}
              value={values.email}
              error={touched.email && errors.email ? true : false}
              helperText={touched.email && errors.email}
            />
          </Grid>
          <Grid xs={12} item>
            <TextField
              id="position"
              label="Position"
              fullWidth
              onChange={handleChange}
              value={values.position}
              error={touched.position && errors.position ? true : false}
              helperText={touched.position && errors.position}
            />
          </Grid>
          <Grid xs={12} item>
            <FormControlLabel
              control={
                <Checkbox
                  id="isActive"
                  onChange={handleChange}
                  checked={values.isActive}
                  value={values.isActive}
                />
              }
              label='Active User'
            />
          </Grid>
          <Grid xs={12} item>
            <Box mt='20px'>
              <Button
                id="create"
                variant="contained"
                color="primary"
                type="submit"
                disabled={isSubmitting}
                fullWidth
              >
                {isUpdate ? 'Update' : 'Create'}
              </Button>
            </Box>
          </Grid>
        </Grid>
      </Box>
    </form>
  );
}

export default User;