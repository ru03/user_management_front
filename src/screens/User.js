import React, { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { Formik } from 'formik';
import { Container } from '@material-ui/core';
import { Alert } from '../components/UI';
import { userValidation } from '../utils/validations'
import UserForm from '../components/forms/users/User';
import useFetch from '../hooks/fetch';
import env from '../config/config';

const User = () => {
  const { data: userCreated, error: userCreatedError, sendRequest: createUser } = useFetch();
  const { clear: clearUpdated, data: userUpdated, error: userUpdatedError, sendRequest: updateUser } = useFetch();
  const { clear, data: user, error: userError, sendRequest: sendUserReq } = useFetch();
  const [initValues, setInitValues] = useState({
    name: '',
    lastName: '',
    email: '',
    position: '',
    isActive: false,
  });
  const [alert, setAlert] = useState({ isSucceeded: false, message: '', isResetForm: false });
  const { id } = useParams();

  useEffect(() => {
    if (userCreatedError || userError) setAlert({ message: userCreatedError || userError, isSucceeded: false, isResetForm: false });
    if (userCreated) {
      setAlert({ message: 'Users was created', isSucceeded: true, isResetForm: true });
    }
    if (user) {
      const { users: { name, lastName, email, active, position } } = user;
      setInitValues({ name, lastName, email, isActive: active !== 0, position });
    };

    return () => {
      setAlert({ isSucceeded: false, message: '', isResetForm: false })
    }
  }, [userCreatedError, userCreated, user, userError]);

  //Update effect
  useEffect(() => {
    if (userUpdatedError) setAlert({ message: userUpdatedError, isSucceeded: false, isResetForm: false });
    if (userUpdated) {
      setAlert({ message: 'Users was updated', isSucceeded: true, isResetForm: false });
    }

    return () => {
      setAlert({ isSucceeded: false, message: '', isResetForm: false })
    }
  }, [clearUpdated, userUpdatedError, userUpdated])

  //Fetch user if exist
  useEffect(() => {
    if (id) {
      const config = {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json'
        },
      }
      sendUserReq(`${env.basepath}${env.users}/${id}`, config);
    };
    if (id === undefined) setInitValues({ name: '', lastName: '', email: '', isActive: false, position: '' });

    return () => {
      clear();
    }
  }, [id, sendUserReq, clear]);

  const onSubmit = (values, helpers) => {
    id ? onUpdateUser(values) : onCreateUser(values);
    helpers.setSubmitting(false);
  };

  const onCreateUser = async (values) => {
    const config = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(values),
    }
    createUser(`${env.basepath}${env.users}`, config);
  }

  const onUpdateUser = async (values) => {
    const config = {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(values),
    }
    updateUser(`${env.basepath}${env.users}/${id}`, config);
  }

  return (
    <Container>
      <Alert message={alert.message} isSucceeded={alert.isSucceeded} />
      <Formik
        initialValues={initValues}
        enableReinitialize
        onSubmit={onSubmit}
        validate={userValidation}
      >
        {props => <UserForm {...props} isResetForm={alert.isResetForm} />}
      </Formik>
    </Container >
  )
};

export default User;