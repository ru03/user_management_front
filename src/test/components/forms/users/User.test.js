import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import User from '../../../../components/forms/users/User';

describe('Login Form Component', () => {
  let props;
  beforeEach(() => {
    props = {
      errors: {},
      handleChange: jest.fn(),
      handleSubmit: jest.fn(),
      isResetForm: false,
      isSubmitting: false,
      isUpdate: false,
      resetForm: jest.fn(),
      touched: {},
      values: {},
    }
  });

  it('mounts', () => {
    const { asFragment } = render(<User {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls handle change on each field', () => {
    const { container, getByPlaceholderText } = render(<User {...props} />)
    fireEvent.change(getByPlaceholderText('Name'), { target: { value: 'value' } });
    fireEvent.change(getByPlaceholderText('Last Name'), { target: { value: 'value' } });
    fireEvent.change(getByPlaceholderText('Email'), { target: { value: 'value' } });
    fireEvent.change(getByPlaceholderText('Position'), { target: { value: 'value' } });
    fireEvent.click(container.querySelector('#isActive'));
    expect(props.handleChange).toHaveBeenCalledTimes(5);
    expect(getByPlaceholderText('Name').value).toBe('value');
    expect(getByPlaceholderText('Last Name').value).toBe('value');
    expect(getByPlaceholderText('Email').value).toBe('value');
    expect(getByPlaceholderText('Position').value).toBe('value');
    expect(container.querySelector('#isActive').checked).toBeTruthy();
  });

  it('shows error name, last name, email and position', () => {
    props = {
      ...props, errors: {
        name: 'Name is required!',
        lastName: 'Last name is required!',
        email: 'Email is required!',
        position: 'Position is required!',
      },
      touched: { name: true, lastName: true, email: true, position: true, }
    }
    const { getByText } = render(<User {...props} />)
    expect(getByText('Name is required!')).toBeInTheDocument();
    expect(getByText('Last name is required!')).toBeInTheDocument();
    expect(getByText('Email is required!')).toBeInTheDocument();
    expect(getByText('Position is required!')).toBeInTheDocument();
  });

  it('calls handleSubmit when click on Login', () => {
    const { getByTestId } = render(<User {...props} />)
    fireEvent.submit(getByTestId('userForm'));
    expect(props.handleSubmit).toHaveBeenCalledTimes(1);
  });

  it('disables create button', () => {
    props = { ...props, isSubmitting: true };
    const { getByTestId } = render(<User {...props} />)
    expect(getByTestId('buttonCreate')).toBeDisabled();
  });

  it('has create button if isUpdate = false', () => {
    const { getByText } = render(<User {...props} />)
    expect(getByText('Create')).toBeInTheDocument();
  });

  it('has update button if isUpdate = true', () => {
    props = { ...props, isUpdate: true };
    const { getByText } = render(<User {...props} />)
    expect(getByText('Update')).toBeInTheDocument();
  });

  it('calls resetForm if isResetForm = true', () => {
    props = { ...props, isResetForm: true };
    render(<User {...props} />);
    expect(props.resetForm).toHaveBeenCalledTimes(1);
  });
});