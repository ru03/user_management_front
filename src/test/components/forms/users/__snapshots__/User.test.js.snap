// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form Component mounts 1`] = `
<DocumentFragment>
  <form
    data-testid="userForm"
  >
    <div
      class="MuiBox-root MuiBox-root-2"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-align-content-xs-center"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
              for="name"
              id="name-label"
            >
              Name
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="name"
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
              for="lastName"
              id="lastName-label"
            >
              Last Name
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="lastName"
                placeholder="Last Name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
              for="email"
              id="email-label"
            >
              Email
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="email"
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
              for="position"
              id="position-label"
            >
              Position
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="position"
                placeholder="Position"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-177 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
              data-testid="activeUser"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-180"
                  data-indeterminate="false"
                  id="isActive"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Active User
            </span>
          </label>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiBox-root MuiBox-root-232"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
              data-testid="buttonCreate"
              id="create"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                Create
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
